docs/design/2018-08-29-new-planner.md

```diff
@@ -0,0 +1,411 @@
+
+### optimization phase: implementation
+
+The target of this phase is searching the best physical plan for a `Group` which
+satisfies the physical property that the parent operator requires.
+
+In this phase, we need to enumerate all the applicable implementation rules for
+each expression in each group under the required physical property. A memo
+structure is used for a group to reduce the repeated search on the same required
+physical property. The search engine used in the new planner can be expressed to
+the pseudo code:
+
+```go
+func implGroup(g *Group, reqPhysProp *PhysicalProperty, costLimit float64) (groupImpl PhysicalPlan) {
+	if g.implemented(reqPhysProp) {
+		return g.getImplementation(reqPhysProp)
+	}
+
+	// handle implementation rules for each equivalent expression.
+    ...
+
+	// handle enforcing rules for the required physical property.
+    ...
+
+	g.insertImpl(reqPhysProp, groupImpl)
+	return groupImpl
+}
+```
+
+In order to find the best physical implementation for the group, we need to
+enumerate all the possible implementations for each group expression under the
+required physical property, the procedure to handle the implementation rules for
+each equivalent expression is:
+
+```go
+// handle implementation rules for each equivalent expression.
+for _, curExpr := range g.equivalents {
+	for _, impl := range implGroupExpr(curExpr) {
+		impl.setCumCost(impl.getSelfCost())
+		for i, childGroup := range curExpr.children {
+			childImpl := implGroup(childGroup, impl.getChildReqProp(i), costLimit-impl.getCumCost())
+			if childImpl == nil {
+				cumCost = math.MaxFloat64
+				break
+			}
+			impl.setChild(i, childImpl)
+			impl.setCumCost(impl.getCumCost() + childImpl.getCumCost())
+		}
+		if groupImpl.getCumCost() > impl.getCumCost() {
+			groupImpl = impl
+		}
+	}
+}
+```
+
+To enumerate all the implementation for an expression, we have to enumerate all
+the applicable implementation rules on that expression, and calculate the
+self-cost for physical implementation:
+
+```go
+func implGroupExpr(cur *GroupExpr, reqPhysProp *PhysicalProperty) (impls []PhysicalPlan) {
+	for _, rule := range GetImplementationRules(cur.exprNode) {
+		if !rule.match(cur, reqPhysProp) {
+			continue
+		}
+		impl := rule.onImplement(cur)
+		impl.calcSelfCost(cur.children)
+		impls = append(impls, impl)
+	}
+	return impls
+}
+```
+
+Also, the enforcing rules should be considered:
+
+```go
+// handle enforcing rules for the required physical property.
+for _, rule := range getEnforcerRules(reqPhysProp) {
+	impl, newReqPhysProp := rule.onEnforce(reqPhysProp)
+	impl.calcSelfCost(g)
+	impl.setCumCost(impl.getSelfCost())
+	childImpl := implGroup(g, newReqPhysProp, costLimit-impl.getCumCost())
+	impl.setChild(0, childImpl)
+	impl.setCumCost(impl.getCumCost() + childImpl.getCumCost())
+	if groupImpl.getCumCost() > impl.getCumCost() {
+		groupImpl = impl
+	}
+}
+```
+
+## Rationale
+
+## Compatibility
+
+No
+
+## Implementation
```

Comment:
Next time, you can use an ordered list (1,2,3,4,5) to describe a procedure. Then your doc will be more readable~

Reply:
done!